src/components/Modals/NewCourse/CourseSelector.vue

```diff
@@ -0,0 +1,156 @@
+import Vue from 'vue';
+import coursesJSON from '@/assets/courses/courses.json';
+
+import { CourseJson } from '@/requirements/courses-json-generator';
+
+export type MatchingCourseSearchResult = {
+  readonly title: string;
+  readonly roster: string;
+  readonly id: number;
+};
+
+const getMatchingCourses = (
+  searchText: string,
+  courses: CourseJson
+): readonly MatchingCourseSearchResult[] => {
+  // search after value length of 2 to reduce search times of courses
+  if (!searchText || searchText.length < 2) return [];
+  /* code array for results that contain course code and title array for results that contain title */
+  const code: MatchingCourseSearchResult[] = [];
+  const title: MatchingCourseSearchResult[] = [];
+
+  for (const attr in courses) {
+    if (courses[attr]) {
+      const result = {
+        title: `${attr}: ${courses[attr].t}`,
+        roster: courses[attr].r,
+        id: courses[attr].i,
+      };
+      if (attr.toUpperCase().includes(searchText) && attr !== 'lastScanned') {
+        code.push(result);
+      } else if (courses[attr].t && courses[attr].t.toUpperCase().includes(searchText)) {
+        title.push(result);
+      }
+    }
+  }
+
+  // Sort both results by title
+  code.sort((first, second) => first.title.localeCompare(second.title));
+  title.sort((first, second) => first.title.localeCompare(second.title));
+
+  /* prioritize code matches over title matches */
+  // limit the number of results to 10
+  return code.concat(title).slice(0, 10);
+};
+
+export default Vue.extend({
+  props: { searchBoxClassName: String, placeholder: String, autoFocus: Boolean },
+  data() {
+    return {
+      searchText: '',
+      currentFocus: -1,
+    };
+  },
+  computed: {
+    matches(): readonly MatchingCourseSearchResult[] {
+      return getMatchingCourses(this.searchText.toUpperCase(), coursesJSON);
+    },
+  },
+  mounted() {
+    // Activate focus and set input to empty
+    const input = this.dropdownInputRef();
+    if (this.autoFocus) input.focus();
+  },
+  methods: {
+    dropdownInputRef(): HTMLInputElement {
+      return this.$refs.dropdownInput as HTMLInputElement;
+    },
+    onEscape() {
+      this.$emit('on-escape');
+    },
+    selectCourse(result: MatchingCourseSearchResult) {
+      this.$emit('on-select', result);
+      this.searchText = result.title;
+      this.currentFocus = -1;
+    },
+    onEnter() {
+      const result = this.matches[this.currentFocus];
+      if (result != null) this.selectCourse(result);
+    },
+  },
+});
+</script>
+
+<style lang="scss">
+@import '@/assets/scss/_variables.scss';
+
+.search-box {
+  border: 1px solid transparent;
+  background-color: #f1f1f1;
+  padding: 10px;
+  font-size: 16px;
+}
+.autocomplete {
+  /*the container must be positioned relative:*/
+  position: relative;
+  display: inline-block;
+  width: 100%;
+  margin-top: 0.5rem;
+  padding-bottom: 12px;
+}
+.autocomplete-items {
+  position: absolute;
+  border: 1px solid #d4d4d4;
```

Comment:
Can we have scss variables for these colors?

Reply:
added